### PR TITLE
Migrate the Animations Package to NNBD

### PR DESCRIPTION
**Pull request by pennzht**
This PR migrates the `animations` package to NNBD.

Thank you very much!

**Currently blocked on:**
* Linter rule `avoid_as`: cannot find good alternative. https://dart-lang.github.io/linter/lints/avoid_as.html

### CODE REVIEW

File: packages/animations/example/lib/container_transition.dart
Comment (JoseAlba) [inline packages/animations/example/lib/container_transition.dart:292]: nit: why is this necessary?

File: packages/animations/example/lib/main.dart
Comment (JoseAlba) [inline packages/animations/example/lib/main.dart:77]: Is there a push to make Navigator.of(context) nnbd similar to Theme.of(context). ?

File: packages/animations/lib/src/open_container.dart
Comment (JoseAlba) [inline packages/animations/lib/src/open_container.dart:77]: What does T extends Object? do for the generic. Does this mean that T can be a class that can be null ?

File: packages/animations/lib/src/page_transition_switcher.dart
Comment (JoseAlba) [inline packages/animations/lib/src/page_transition_switcher.dart:38]: Remove this comment

Comment (pennzht) [reply packages/animations/lib/src/page_transition_switcher.dart]: Done.

File: packages/animations/example/lib/main.dart
Comment (pennzht) [reply packages/animations/example/lib/main.dart:77]: Navigator.of(context) is defined to return a `NavigatorState?` inside flutter/flutter, so we can't do anything about it.

File: packages/animations/lib/src/open_container.dart
Comment (pennzht) [reply packages/animations/lib/src/open_container.dart:79]: I tried removing this, but the analyzer complained with:

```
  error • Missing type arguments for generic type 'OpenContainer<dynamic>' at example/lib/container_transition.dart:241:29 • (implicit_dynamic_type)
  error • Missing type arguments for generic type 'OpenContainer<dynamic>' at test/open_container_test.dart:24:18 • (implicit_dynamic_type)
  error • Missing type arguments for generic type 'OpenContainer<dynamic>' at test/open_container_test.dart:197:18 • (implicit_dynamic_type)
  error • Missing type arguments for generic type 'OpenContainer<dynamic>' at test/open_container_test.dart:366:16 • (implicit_dynamic_type)
  error • Missing type arguments for generic type 'OpenContainer<dynamic>' at test/open_container_test.dart:541:16 • (implicit_dynamic_type)
  error • Missing type arguments for generic type 'OpenContainer<dynamic>' at test/open_container_test.dart:715:16 • (implicit_dynamic_type)
  error • Missing type arguments for generic type 'OpenContainer<dynamic>' at test/open_container_test.dart:739:16 • (implicit_dynamic_type)
...
```

File: packages/animations/lib/src/dual_transition_builder.dart
Comment (goderbauer) [inline packages/animations/lib/src/dual_transition_builder.dart:19]: To avoid that this one and the one in the framework diverge: Can you just remove this file all together and use the DualTransitionBuilder from flutter/flutter (see TODO at top of this file)?

File: packages/animations/test/fade_scale_transition_test.dart
Comment (goderbauer) [inline packages/animations/test/fade_scale_transition_test.dart:5]: Why? Please leave a comment per flutter style guide.

Comment (goderbauer) [reply packages/animations/test/fade_scale_transition_test.dart:5]: (we may just want to turn this lint off in the analyzer options)

Comment (goderbauer) [reply packages/animations/test/fade_scale_transition_test.dart:5]: Same in other files.

File: packages/animations/lib/src/open_container.dart
Comment (goderbauer) [inline packages/animations/lib/src/open_container.dart:275]: this formatting is very strange, is that what dartfmt produces?

File: packages/animations/example/lib/fade_scale_transition.dart
Comment (gspencergoog) [reply packages/animations/example/lib/fade_scale_transition.dart:123]: Oh! I thought we already took care of that one... I'll put together a PR.

Comment (gspencergoog) [reply packages/animations/example/lib/fade_scale_transition.dart:123]: See https://github.com/flutter/flutter/pull/70726

**Comment by shihaohong** [comment]
We discussed this in an email, but another blocker is that since the CI runs on the stable version of Flutter, this PR is blocked by that as well.

File: packages/animations/lib/src/dual_transition_builder.dart
Comment (pennzht) [reply packages/animations/lib/src/dual_transition_builder.dart]: Done.

File: packages/animations/test/fade_scale_transition_test.dart
Comment (pennzht) [reply packages/animations/test/fade_scale_transition_test.dart:5]: How can we turn this lint off? I tried adding an `analyzer_options.yaml` in `packages/animations/`, but it didn't work.

**Should we edit the `analyzer_options.yaml` file in the project root directory?**

File: packages/animations/lib/src/open_container.dart
Comment (pennzht) [reply packages/animations/lib/src/open_container.dart:275]: This is indeed produced by `flutter format .`

File: packages/animations/example/lib/container_transition.dart
Comment (pennzht) [reply packages/animations/example/lib/container_transition.dart]: Modified type signature to remove this.

**Comment by pennzht** [comment]
I still cannot make the ["publishable" check](https://github.com/flutter/packages/pull/245/checks?check_run_id=1418786685) pass. What should I do?

File: packages/animations/lib/src/open_container.dart
Comment (JoseAlba) [inline packages/animations/lib/src/open_container.dart:6]: empty line not used. Plus the space above is using a linter which I don't think would apply to anything since line 6 has nothing

File: packages/animations/lib/src/open_container.dart
Comment (JoseAlba) [inline packages/animations/lib/src/open_container.dart:679]: Since we are not doing anything with the null case I think it makes sense to drop this case statement.

File: packages/animations/lib/src/open_container.dart
Comment (JoseAlba) [inline packages/animations/lib/src/open_container.dart:792]: nit: Can the begin variable be null. Maybe it makes sense to make begin end NNBD that way we don't have to use the ! all the time.

File: packages/animations/lib/src/page_transition_switcher.dart
Comment (JoseAlba) [inline packages/animations/lib/src/page_transition_switcher.dart:191]: Should there be a ? for child ? Can this class have a null child ?

File: packages/animations/lib/src/open_container.dart
Comment (pennzht) [reply packages/animations/lib/src/open_container.dart:6]: This comment asks the analyzer to ignore `avoid_as` throughout this file.

File: packages/animations/lib/src/open_container.dart
Comment (pennzht) [reply packages/animations/lib/src/open_container.dart:679]: Dart analyzer gives this warning when `case null` is removed:

```
warning • Missing case clause for 'null' at open_container.dart:682:5 • (missing_enum_constant_in_switch)
```

`case null` is equivalent to `default` here.

File: packages/animations/lib/src/open_container.dart
Comment (pennzht) [reply packages/animations/lib/src/open_container.dart:792]: `RectTween` is defined by the flutter framework as follows, so I can't really change it.

```dart
/// An interpolation between two rectangles.
///
/// This class specializes the interpolation of [Tween<Rect>] to use
/// [Rect.lerp].
///
/// The values can be null, representing a zero-sized rectangle at the
/// origin ([Rect.zero]).
///
/// See [Tween] for a discussion on how to use interpolation objects.
class RectTween extends Tween<Rect?> {
  /// Creates a [Rect] tween.
  ///
  /// The [begin] and [end] properties may be null; the null value
  /// is treated as an empty rect at the top left corner.
  RectTween({ Rect? begin, Rect? end }) : super(begin: begin, end: end);

  /// Returns the value this variable has at the given animation clock value.
  @override
  Rect? lerp(double t) => Rect.lerp(begin, end, t);
}
```

File: packages/animations/lib/src/page_transition_switcher.dart
Comment (pennzht) [reply packages/animations/lib/src/page_transition_switcher.dart:191]: There should be a `?` for child. This class can have a null child, a case which is handled on L307:
https://github.com/flutter/packages/blob/master/packages/animations/lib/src/page_transition_switcher.dart#L307

File: packages/animations/lib/src/open_container.dart
Comment (pennzht) [reply packages/animations/lib/src/open_container.dart:6]: We're still looking for better ways to ignore `avoid_as`.

File: packages/animations/test/fade_scale_transition_test.dart
Comment (goderbauer) [reply packages/animations/test/fade_scale_transition_test.dart:5]: It should be fine to turn this off in the project root directory.

Comment (pennzht) [reply packages/animations/test/fade_scale_transition_test.dart]: Done. Thank you!

File: packages/animations/lib/src/open_container.dart
Comment (pennzht) [reply packages/animations/lib/src/open_container.dart]: Done; please see comment above.

**Comment by pennzht** [comment]
@goderbauer @gspencergoog @shihaohong 

Hi! How is everything going?

I'm just wondering:
* When can we merge this?
* When can we publish this?

**Comment by shihaohong** [comment]
I think Michael and Greg are out until next week (it's Thankgiving week in the US). I don't think either of them commented on publishing and merging the NNBD version of the animations package and while I'm on board, I want to make sure we iron out any versioning weirdness that could happen on GitHub and pub. 

So far, I've come up with two scenarios that could occur, but neither seem to raise any serious red flags for me:
1. NNBD prerelease is published (`2.0.0-nullsafety.0`).
2. Hypothetically, some new `packages/animations` features are added and someone unknowingly tries to publish the next version before NNBD reaches `flutter/stable`:
- Publishing as `1.1.2+x`: On `packages/master` we're on `1.1.2` and someone tries publishing new features/fixes on `master`. This probably will not cause any issues, but I'm not sure what would actually happen (like pub disallowing publishing versions older than the latest version), since I've never tried this.
- Publishing as `2.0.0-nullsafety.0+x`: Someone could unknowingly try to publish a `2.0.0-nullsafety.0+x` from `packages/master`, which would be bad since it wouldn't actually contain the NNBD migration. However, we could just publish another prerelease and increment the prerelease version number on top of it to correct this.

So basically, if we do make publish a prerelease, we need to be clear that new PRs for `packages/animations` are not incorrectly published.

## Short Term (before `flutter/stable` has NNBD)
1. We're merging the NNBD migration to a `packages/nnbd` branch instead of `packages/master`. 
2. We'll publish `packages/animations` from `nnbd` as `2.0.0-nullsafety.x`. Any unexpected bugs in the NNBD version should be fixed in the `nnbd` branch.

## Longer Term (once `flutter/stable` is NNBD)
1. Make sure `packages/animations` is not published beyond `2.0.0-nullsafety.0` from the `master` branch until the NNBD changes are properly merged into `packages/animations` on the `master` branch.
2. Once `flutter/flutter` is NNBD stable, we merge the `nnbd` branch into the `master` branch and publish a stable `2.0.0` release.

cc/ @HansMuller, who might be able to provide some feedback while Greg and Michael are out on what he thinks about this process.

File: packages/animations/lib/src/fade_scale_transition.dart
Comment (shihaohong) [inline packages/animations/lib/src/fade_scale_transition.dart:8]: Nice, thanks for removing this!

**Review by shihaohong — APPROVED** [review]

**Comment by pennzht** [comment]
@shihaohong 

Thank you for your message! Overall this sounds good to me, as long as we can publish `1.1.2+x` after we publish `2.0.0-nullsafety.0`

The 2-branch model is currently being used in some packages such as `crypto` (see https://github.com/dart-lang/crypto/tree/3.0 for a null-safe branch).

I think it's fine to publish newer prerelease versions for the null-safe branch, such as `2.0.0-nullsafety.1`, as long as it does not reach `2.0.0`

Also, if `animations` is a frequently published package, it's probably a good idea to keep the main and null-safe branches in sync regularly.

**Comment by shihaohong** [comment]
We don't publish `animations` very frequently, and the number of PRs we've been getting have been pretty low so I think we should be good with the plan as long as we're proactive with merging `nnbd` to `master` once `flutter/stable` reaches NNBD :)

**Comment by shihaohong** [comment]
@pennzht `crypto`'s `3.0` null-safe branch isn't published, even as a prerelease though right? Would that then be the main difference between `crypto` and `animations`

**Comment by pennzht** [comment]
@shihaohong 

> @pennzht `crypto`'s `3.0` null-safe branch isn't published right, even as a prerelease? Would that be the main difference between how we're handling `crypto` and `animations`?

Correct, `crypto`'s null-safe version has not been published yet, even as a prerelease.

**Publishing** `animations` is not very urgent at the moment; your comment https://github.com/flutter/packages/pull/245#issuecomment-732642312 sounds good to me.

**Having a migrated version available**, even in a separate `nnbd` branch, is rather urgent, since we need to unblock [Motion Codelab](https://github.com/material-components/material-components-flutter-motion-codelab) and [Gallery](https://github.com/flutter/gallery), both of which depend on `animations`.
Even if `animations` is not published, we can still use a git reference for the moment.

**Would it be fine to merge this PR?** The CI is currently [blocking](https://github.com/flutter/packages/pull/245/checks?check_run_id=1420554635), but I don't know how to resolve it.

Thank you!

**Comment by pennzht** [comment]
@rami-a

**Comment by shihaohong** [comment]
I just figured it out! You need to update the Dockerfile to use the `beta-web` version of Flutter for the CI checked in the `nnbd` branch. Here is [an example](https://github.com/flutter/packages/pull/248/files#diff-fd0c8401badda82156f9e7bd621fa3a0e586d8128e4a80af17c7cbff70bee11eR1). 

This will likely cause some NNBD-related errors that you'll need to fix -- Once you've made the Dockerfile change and fixed any remaining NNBD issues, I'll review the PR again :) I'll also edit the existing comment above to include making sure we use `stable-web` when we merge the `package/nnbd` and `package/stable` when `flutter/stable` reaches NNBD.

**Comment by pennzht** [comment]
@shihaohong 

Thank you so much for helping us on this migration! Please let us know how it goes.
Also, if you need anything from us, please let us know too.

**Comment by shihaohong** [comment]
The CI modifications were made in https://github.com/flutter/packages/pull/249 and we should be able to unblock the CI checks once this branch has been merged with the `nnbd` branch.

File: packages/animations/example/lib/shared_axis_transition.dart
Comment (goderbauer) [inline packages/animations/example/lib/shared_axis_transition.dart:17]: It is odd that this field is nullable.

Comment (rami-a) [reply packages/animations/example/lib/shared_axis_transition.dart:17]: It has to be nullable since the radio button's `onChanged` callback is `final ValueChanged<T?>? onChanged`

**Comment by rami-a** [comment]
Closing this since https://github.com/flutter/packages/pull/253 was merged and is the same thing

**Pull request closed without merging**
